File: speed-cpp/src/AccessRegistry.cpp
#include "../include/AccessRegistry.hpp"

namespace SPEED {

AccessRegistry::AccessRegistry(const std::filesystem::path &ac_path,
                               const std::string &proc_name)
    : ac_path_(ac_path), access_filename_(proc_name) {
  if (!Utils::directoryExists(ac_path)) {
    Utils::createAccessRegistryDir(ac_path);
  }
  if (Utils::fileExists(ac_path_ / (proc_name + ".oregistry"))) {
    removeAccessFile();
  }
  this->proc_name_ = proc_name;
  putAccessFile();
  incrementalBuildGlobalRegistry();
  printRegistry();
}

const std::filesystem::path &AccessRegistry::getAccessRegistryPath() const {
  return ac_path_;
}

void AccessRegistry::incrementalBuildGlobalRegistry() {
  try {
    for (const auto &entry :
         std::filesystem::recursive_directory_iterator(ac_path_)) {
      std::string name = entry.path().stem().string();
      if (name == proc_name_)
        continue;
      if (global_registry_.find(name) == global_registry_.end()) {
        global_registry_.insert(name);
        std::cout << "[INFO] Added to registry: " << name << "\n";
      } else {
        std::cout << "[SKIP] Already present: " << name << "\n";
      }
    }
  } catch (const std::filesystem::filesystem_error &e) {
    std::cerr << "[ERROR] Filesystem error: " << e.what() << "\n";
  }
}
const std::unordered_set<std::string>
AccessRegistry::getGlobalRegistry() const {
  return global_registry_;
}
const std::unordered_set<std::string> AccessRegistry::getAccessList() const {
  return allowedProcesses_;
}
const std::unordered_set<std::string> AccessRegistry::getConnectedList() const {
  return connected_list_;
}
void AccessRegistry::printRegistry() const {
  std::cout << "=== Global Registry ===\n";
  for (const auto &item : global_registry_) {
    std::cout << item << "\n";
  }
}

void AccessRegistry::putAccessFile() {
  std::lock_guard<std::mutex> lock(mtx_);
  const std::filesystem::path before_path =
      ac_path_ / (proc_name_ + ".iregistry");
  const std::filesystem::path after_path =
      ac_path_ / (proc_name_ + ".oregistry");
  std::ofstream outstream(before_path);
  outstream << proc_name_ << "\n";
  outstream.close();
  std::rename(before_path.c_str(), after_path.c_str());
}
bool AccessRegistry::checkGlobalRegistry(const std::string &proc_name) const {
  return global_registry_.find(proc_name) != global_registry_.end();
}

bool AccessRegistry::checkAccess(const std::string &proc_name) const {
  return allowedProcesses_.find(proc_name) != allowedProcesses_.end();
}

void AccessRegistry::addProcessToList(const std::string &proc_name) {
  // std::cout << "Inside addProcessToList\n";
  std::lock_guard<std::mutex> lock(mtx_);

  if (checkAccess(proc_name)) {
    std::cout << "[ERROR]: Process Already Exists in Access Registry\n";
    return;
  }
  // Safe add
  allowedProcesses_.insert(proc_name);
  // connect_to(proc_name);
}
bool AccessRegistry::connect_to(const std::string &proc_name) {
  if (connected_list_.find(proc_name) != connected_list_.end()) {
    return false;
  }
  connected_list_.insert(proc_name);
  return true;
}
void AccessRegistry::syncAccessRegistry() {
  for (const auto &entry : global_registry_) {
    if (!checkAccess(entry)) {
      allowedProcesses_.erase(entry);
    }
  }
}
bool AccessRegistry::removeProcessFromGlobalRegistry(
    const std::string &proc_name) {
  std::lock_guard<std::mutex> lock(mtx_);
  auto it = global_registry_.find(proc_name);
  if (it != global_registry_.end()) {
    global_registry_.erase(it);
    std::cout << "[INFO]: Process removed from Global Registry\n";
    return true;
  }
  std::cout << "[WARN]: Process not in Global Registry\n";
  return false;
}
bool AccessRegistry::removeProcessFromAccessList(const std::string &proc_name) {
  std::lock_guard<std::mutex> lock(mtx_);

  auto it = allowedProcesses_.find(proc_name);
  if (it != allowedProcesses_.end()) {
    allowedProcesses_.erase(it);
    std::cout << "[INFO]: Process removed from Allowed Registry\n";
    return true;
  }
  std::cout << "[WARN]: Process not in Allowed Registry\n";
  return false;
}
bool AccessRegistry::removeProcessFromConnectedList(
    const std::string &proc_name) {
  auto it = connected_list_.find(proc_name);
  if (it != connected_list_.end()) {
    connected_list_.erase(it);
    std::cout << "[INFO]: Process removed from Connected List\n";
    return true;
  }
  std::cout << "[WARN]: Process not in Connected List\n";
  return false;
}
void AccessRegistry::removeAccessFile() {
  std::lock_guard<std::mutex> lock(mtx_);
  const std::filesystem::path ac_removal_path_ =
      ac_path_ / (proc_name_ + ".oregistry");
  if (!Utils::fileExists(ac_removal_path_)) {
    std::cout
        << "[ERROR]: Unable to remove the access file, file doesnt exist\n";
  }
  std::filesystem::remove(ac_removal_path_);
}
bool AccessRegistry::check_connection(const std::string &proc_name) const {
  std::lock_guard<std::mutex> lock(mtx_);
  if (connected_list_.find(proc_name) == connected_list_.end())
    return false;

  return true;
}
void AccessRegistry::try_connect_all() {
  for (const std::string &proc_name : allowedProcesses_) {
    connect_to(proc_name);
  }
}
} // namespace SPEED

File: speed-python/speed/access_registry.py
import threading
from pathlib import Path
from typing import Set, List
from .utils import Utils

class AccessRegistry:
    def __init__(self, speed_dir: Path, proc_name: str):
        self.speed_dir = speed_dir
        self.proc_name = proc_name
        self.access_registry_dir = speed_dir / "access_registry"
        self.allowed_processes: Set[str] = set()
        self.global_registry: Set[str] = set()
        self.connected_list: Set[str] = set()
        self.mutex = threading.Lock()
        
        # Initialize directories
        Utils.create_default_dir(speed_dir)
        Utils.create_access_registry_dir(self.access_registry_dir)
        
        # Build initial registry
        self.incremental_build_global_registry()
    
    def add_process_to_list(self, proc_name: str):
        """Add process to allowed list"""
        with self.mutex:
            self.allowed_processes.add(proc_name)
    
    def incremental_build_global_registry(self):
        """Build global registry from access registry files"""
        with self.mutex:
            self.global_registry.clear()
            
            if not self.access_registry_dir.exists():
                return
            
            for file_path in self.access_registry_dir.glob("*.oregistry"):
                try:
                    with open(file_path, 'r') as f:
                        proc_name = f.read().strip()
                        self.global_registry.add(proc_name)
                except Exception:
                    continue
    
    def remove_access_file(self):
        """Remove this process's access file"""
        access_file = self.access_registry_dir / f"{self.proc_name}.oregistry"
        if access_file.exists():
            access_file.unlink()
    
    def sync_access_registry(self):
        """Sync access registry with filesystem"""
        # Create/update our access file
        access_file = self.access_registry_dir / f"{self.proc_name}.iregistry"
        try:
            with open(access_file, 'w') as f:
                f.write(self.proc_name)
            # Rename to indicate it's ready
            ready_file = self.access_registry_dir / f"{self.proc_name}.oregistry"
            access_file.rename(ready_file)
        except Exception as e:
            print(f"Error syncing access registry: {e}")
    
    def remove_process_from_global_registry(self, proc_name: str) -> bool:
        """Remove process from global registry"""
        with self.mutex:
            if proc_name in self.global_registry:
                self.global_registry.remove(proc_name)
                return True
            return False
    
    def check_access(self, proc_name: str) -> bool:
        """Check if process is in allowed list"""
        with self.mutex:
            return proc_name in self.allowed_processes
    
    def remove_process_from_access_list(self, proc_name: str) -> bool:
        """Remove process from access list"""
        with self.mutex:
            if proc_name in self.allowed_processes:
                self.allowed_processes.remove(proc_name)
                return True
            return False
    
    def remove_process_from_connected_list(self, proc_name: str) -> bool:
        """Remove process from connected list"""
        with self.mutex:
            if proc_name in self.connected_list:
                self.connected_list.remove(proc_name)
                return True
            return False
    
    def connect_to(self, proc_name: str) -> bool:
        """Add process to connected list"""
        with self.mutex:
            if (proc_name in self.allowed_processes and 
                proc_name in self.global_registry):
                self.connected_list.add(proc_name)
                return True
            return False
    
    def check_global_registry(self, proc_name: str) -> bool:
        """Check if process is in global registry"""
        with self.mutex:
            return proc_name in self.global_registry
    
    def check_connection(self, proc_name: str) -> bool:
        """Check if process is connected"""
        with self.mutex:
            return proc_name in self.connected_list
    
    def get_global_registry(self) -> Set[str]:
        """Get copy of global registry"""
        with self.mutex:
            return self.global_registry.copy()
    
    def get_access_list(self) -> Set[str]:
        """Get copy of access list"""
        with self.mutex:
            return self.allowed_processes.copy()
    
    def get_connected_list(self) -> Set[str]:
        """Get copy of connected list"""
        with self.mutex:
            return self.connected_list.copy()